00_basics/sc_dimensions.ipynb

```diff
@@ -0,0 +1,771 @@
+{
```

Comment:
<pre class="ql-syntax" spellcheck="false">(
Ndef(\x, {
&nbsp;&nbsp; &nbsp;var a, b, mix, selected, gater;
&nbsp;&nbsp; &nbsp;
&nbsp;&nbsp; &nbsp;gater = {
&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; &nbsp;Trig1.kr(
&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; &nbsp;in: PulseDivider.kr(
&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; &nbsp;trig: Impulse.kr(8),
&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; &nbsp;div: IRand(4, 9),
&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; &nbsp;start: IRand(0, 8)),
&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; &nbsp;dur: 1/8
&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; &nbsp;)
&nbsp;&nbsp; &nbsp;};
&nbsp;&nbsp; &nbsp;
&nbsp;&nbsp; &nbsp;// a 4 x 3 matrix of signals
&nbsp;&nbsp; &nbsp;a = [
&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; &nbsp;SinOsc.ar(freq: [440, 226, 782]) * gater.value,
&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; &nbsp;BPF.ar(PinkNoise.ar, freq: [145, 881, 9000], rq: 0.1, mul: 100) * gater.value,
&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; &nbsp;Pulse.ar(freq: [8000, 1000, 500]) * gater.value,
&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; &nbsp;Blip.ar(freq: [3, 8, 20], numharm: 140) * gater.value,
&nbsp;&nbsp; &nbsp;];
&nbsp;&nbsp; &nbsp;
&nbsp;&nbsp; &nbsp;
&nbsp;&nbsp; &nbsp;// calculate a mixdown
&nbsp;&nbsp; &nbsp;// multiplying the matrix of 4 rows with 3 channels each
&nbsp;&nbsp; &nbsp;// with a 3 x 8 matrix will return 4 rows with 8 channels each
&nbsp;&nbsp; &nbsp;b = ((1.0 ! 8) ! 3).rand2.max(0) + 0.1;
&nbsp;&nbsp; &nbsp;b.do { |row| row.round(0.002).postln };
&nbsp;&nbsp;&nbsp; //b = { LFNoise1.kr(1 ! 8).max(0) } ! 3;

&nbsp;&nbsp; &nbsp;mix = ~matrixMul.(a, b);
&nbsp;&nbsp; &nbsp;
&nbsp;&nbsp; &nbsp;mix.shape.postln; // post the dimensions: 4 rows of 8
&nbsp;&nbsp; &nbsp;
&nbsp;&nbsp; &nbsp;// we want each channel to be a mix of the 4 oscillators
&nbsp;&nbsp; &nbsp;// so we sum the 4 rows, and get 8 channels
&nbsp;&nbsp; &nbsp;mix = mix.sum;
&nbsp;&nbsp; &nbsp;
&nbsp;&nbsp; &nbsp;mix.shape.postln; // post the dimensions: array of 8

&nbsp;&nbsp; &nbsp;// tune through the 8 signals with the cursor position
&nbsp;&nbsp; &nbsp;// each has a different mix
&nbsp;&nbsp; &nbsp;selected = SelectX.ar(
&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; &nbsp;which: MouseX.kr(0, 7).round.lag,
&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; &nbsp;array: mix
&nbsp;&nbsp; &nbsp;);
&nbsp;&nbsp; &nbsp;
&nbsp;&nbsp; &nbsp;// expand to stereo
&nbsp;&nbsp; &nbsp;selected ! 2 * 0.1
}).play;
)
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/capital-G/musikinformatik-sose2021/pull/23/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='44'/>

Reply:
This is how t was supposed to work (I've now translated it to row-vector representation)
I'll try and think of a simpler example later.

Reply:
Better example:

```

(
Ndef(\x, {
	var allSounds, allRatios, multiplier, matrixMul, mixes;
	
	
	// array of 5 rows, each with 8 sine tones in a series
	allRatios = [1.2, 1.25, 1.5, pi/2, 5/4];
	allSounds = allRatios.collect { |ratio| 
		8.collect { |i|
			var freq = 220 * (ratio ** i);
			SinOsc.ar(freq) * AmpComp.kr(freq, 220)
		} * LFNoise2.ar(70).max(0)
	};
	allSounds.shape; 
	
	
	// now we would like to make a number of mixes, say 12 different ones
	// each should have a different mix of the 8 sine tones
	// in 5 rows for the different ratios
	
	multiplier = {
		{ 1.0.rand } ! 12  // 12 mixes
	} ! 8; // 8 different amplitudes
	
	matrixMul = { |a, b| 
		if(a[0].size != b.size) { Error("matrix sizes do not match").throw };
		a.collect { |d| (b * d).sum  } 
	};
	mixes = matrixMul.(allSounds, multiplier);
	mixes.shape.postln; // 5 rows, each 12 elements

	// to dynamically select one of the mixes, we use x and y cursor coordinates
	SelectX.ar(
		MouseX.kr(0, 12).trunc.lag,
		SelectX.ar(
			MouseY.kr(0, 5).trunc.lag,
			mixes
		)
	) * 0.2
	
}).play
)

```

Reply:
<p>Trying the 2nd example fails for me with</p><p><code></code><code></p><p>-&gt; Ndef('x')</p><p>late 0.242954502</p><p>exception in GraphDef_Load: exceeded number of interconnect buffers.</p><p>while reading file: '/Users/scheiba/Library/Application Support/SuperCollider/synthdefs/temp__0x1120202026_0.scsyndef'exception in GraphDef_Load: exceeded number of interconnect buffers.</p><p>while reading file: '/Users/scheiba/Library/Application Support/SuperCollider/synthdefs/temp__0x1120202026_0.scsyndef'*** ERROR: SynthDef temp__0x1120202026_0 not found</p><p>FAILURE IN SERVER /s_new SynthDef not found</p><p></code><code></code></p>

Reply:
Ah ok, you need to start the server with more wire bufs: `s.options.numWireBufs = 2048 * 16; s.reboot;` But I can also refactor it a little to make it lighter.

Reply:
I came up with another example based on this which is, i hope, OK